Part 3 of Issue246 block fees and rewards

hewison-chris opened the pull request:
This is the next step to enable calculation of fees for validators who signed the block.

hewison-chris committed:
Store the fees and data_fees for each height

Fees are now only paid to Validators for blocks they sign. The `Coinbase`
transaction is now included in the block at the end of the next payout period.
The fees are the `sum of the inputs` minus the `sum of the outputs`.
If we do not store the fees when we externalise a block then some inputs may be
spent before the payout is made. The data fees are also stored as they are only
paid to the 'Commons Budget' so are subtracted from the fees when paying the
validators their percentage of fees.

Geod24 commented on source/agora/consensus/Ledger.d:
```diff
@@ -608,27 +614,69 @@ public class Ledger
 
         Returns:
-            List of expected Coinbase TXs
+            Coinbase TX
```
`transaction`

Geod24 commented on doc/Database.md:
```diff
@@ -54,14 +54,15 @@ This table should be removed in favor of querying `validator` directly.
+### `block_fees` table
 
-Note sure why we need this.
+Keep track of fees as blocks are externalized (when the UTXO set has all the inputs still)
```
```suggestion
Keep track of fees as blocks are externalized. This needs to be done when a block is externalized,
as once the inputs are consumed, there will be no way to calculate them.
```

A bit more verbose but hopefully clearer too. Alterations welcome.

hewison-chris replied:
Updates

Geod24 commented on source/agora/consensus/Fee.d:
```diff
@@ -353,85 +346,105 @@ public class FeeManager
+
+    public Amount getBlockFees (in Height height) nothrow @safe
+    {
+        if (height !in this.block_fees)
+            return 0.coins;
+        return this.block_fees[height];
```
This is doing a double lookup.
```suggestion
        if (auto ptr = height in this.block_fees)
            return *ptr;
        return 0.coins;
```

Or simply:
```suggestion
        return this.block_fees.get(height, 0.coins);
```

hewison-chris replied:
I like the second one.

Geod24 reviewed (commented):
Why do the data fee have to be in a separate AA as the regular fee ?

Geod24 commented on source/agora/consensus/Ledger.d:
```diff
@@ -1295,22 +1326,24 @@ public class Ledger
+private Height lastPaidHeight(in Height height, uint payout_period) @safe @nogc nothrow pure
 {
-    return Height(height <= payout_period ? 1 : (height - 1) - ((height - 1) % payout_period));
+    // We return 1 before the first payout is made as we use this for block catchup and Genesis is never fetched
```
We don't use it for block catchup AFAICS, only for signature catchup ?

hewison-chris replied:
Correct I will update.

Geod24 commented on source/agora/consensus/Ledger.d:
```diff
@@ -1295,22 +1326,24 @@ public class Ledger
 }
 
-private Height lastPaymentBlock(in Height height, uint payout_period) @safe @nogc nothrow pure
+// This is the last block height that has had fees and rewards paid before the current block
```
```suggestion
/// This is the last block height that has had fees and rewards paid before the current block
```

hewison-chris replied:
Wasn't sure with `private` (or `unittest`) comments whether to use `///` or `//`. Should they also be `///`?

Geod24 replied:
On unittests, use `//` most of the time, unless the unittest is *documenting* the previous function. Most of our unittests are testing the function (hence they are exposing / checking more of the internals), and don't qualify as documenting unittests.

On function, no matter the visibility, prefer `///`.

Geod24 commented on source/agora/consensus/Ledger.d:
```diff
@@ -1295,22 +1326,24 @@ public class Ledger
 }
 
-// expected is the height of last payment block not including current if it is one
+// expected is the height of last block that has had fess and rewards paid
```
fees*

Geod24 commented on source/agora/consensus/Ledger.d:
```diff
@@ -1295,22 +1326,24 @@ public class Ledger
 {
-    return Height(height <= payout_period ? 1 : (height - 1) - ((height - 1) % payout_period));
+    // We return 1 before the first payout is made as we use this for block catchup and Genesis is never fetched
+    return Height(height < 2 * payout_period ? 1 : height - (height % payout_period) - payout_period);
```
I think an `if` here would be more readable than a ternary ?

hewison-chris commented:
> Why do the data fee have to be in a separate AA as the regular fee ?

When we want to check (i.e. is the fee enough to cover the data fee + min fee rate for tx) many blocks later it is convenient to have the data fees stored in their own AA. We could recalculate  but this saves the extra computation and keeps it consistent with the fees storage.

Geod24 commented on source/agora/consensus/Ledger.d:
```diff
@@ -1260,10 +1291,10 @@ public class Ledger
-    /// return the last payment block before the current block
-    public Height getLastPaymentBlock () const scope @safe @nogc nothrow pure
+    /// return the last paid out block before the current block
+    public Height getLastPaidHeight () const scope @safe @nogc nothrow pure
     {
-        return lastPaymentBlock(this.getBlockHeight, this.params.PayoutPeriod);
+        return lastPaidHeight(this.getBlockHeight, this.params.PayoutPeriod);
```
The rename is welcome, but it clutters the diff and makes reviewing harder, that's why having it in its own commit would be best. My litmus test for such a commit separation is: "Will this feature work without this diff?".

hewison-chris replied:
I'll split it out into a separate commit.

Geod24 commented:
> When we want to check (i.e. is the fee enough to cover the data fee + min fee rate for tx) many blocks later [...]

I'm not sure I follow. That check is performed on tx validation, at block externalization, isn't it ?

hewison-chris committed:
Rename `lastPaymentBlock` to `lastPaidHeight`

This is in preperation for when the payment block is no longer the same block
as the last block paid for.

hewison-chris commented:
> > When we want to check (i.e. is the fee enough to cover the data fee + min fee rate for tx) many blocks later [...]
> 
> I'm not sure I follow. That check is performed on tx validation, at block externalization, isn't it ?

We only pay the data fees to the `Commons Budget` so we deduct them from the total fees before paying the validators the `ValidatorTXFeeCut` percentage. I updated the commit message to include more details about this.

hewison-chris commented on source/agora/consensus/Ledger.d:
```diff
@@ -1306,11 +1307,11 @@ unittest
-    only(tuple(0,1), tuple(1,1), tuple(4,1), tuple(5,1), // test before first reward block is externalized
-        tuple(6,5), tuple(9,5), tuple(10,5), // test before second reward block is externalized
-        tuple(11,10)).each!( // last we test after second
-        (height, expected) => assert(lastPaymentBlock(Height(height), paymentPeriod) == expected));
```
I will fix this

hewison-chris replied:
Fixed

hewison-chris commented on source/agora/consensus/Ledger.d:
```diff
@@ -1260,8 +1292,8 @@ public class Ledger
-    /// return the last payment block before the current block
-    public Height getLastPaymentBlock () const scope @safe @nogc nothrow pure
+    /// return the last paid out block before the current block
+    public Height getLastPaidHeight () const scope @safe @nogc nothrow pure
```
Missed this rename. Will update.

hewison-chris replied:
Moved into rename commit

hewison-chris commented:
I still have some updates to some new code in this PR before this is merged which will make the next PR have less changes.

hewison-chris committed:
Transaction Fee: Improve check function

It missed the check that the included fees were enough to cover the data fee and
the minimum tx fee calculated from the minimum fee per byte.
Also the function documentation was updated.

hewison-chris commented:
Updated and green :-)

Geod24 reviewed (approved):
LGTM. A few minors items but we can address those later. Thanks for the extended commit msg.
